Add back in the draft migration information (#635)
Qiskit/qiskit-ibm-runtime: docs/migrate/migrate-e2e.rst

End to end example
==================

As a comprehensive example, we will walk through a common use case that
would require the user to migrate code to use the Qiskit Runtime
execution model. 

Overview
--------

Recall that when choosing which primitive to use, we first need to
understand whether our algorithm needs to generate the entire quasi-probability
distribution, or expectation values of observables. 

In our use case, an iterative amplitude estimation (IAE), we want to
discover one of the probability amplitudes of a quantum state. Thus, the
algorithm should rely on the Sampler primitive.

.. _amplitude:

Amplitude estimation use case
-----------------------------

In this example, we look at the case that uses IAE approach to solve an
estimation problem. Historically, on IBM Quantum Platform, users could
achieve this by configuring epsilon and alpha parameters on the

> ``QuantumInstance`` object, as shown in the code example without

Reviewer: But the code doesn't pass epsilon and alpha to `QuantumInstance`. They are passed to `IterativeAmplitudeEstimation`

       'epsilon_estimated_processed': 0.0006248869813553215,
       'epsilon_target': None,
       'estimate_intervals': [   [0.0, 1.0],
                                 [0.2828417741377001, 0.3216648682822661],
                                 [0.299362359799064, 0.30061213376177465]],
       'estimation': 0.2999872467804193,
       'estimation_processed': 0.2999872467804193,
       'num_oracle_queries': 61440,
       'powers': [0, 0, 15],
       'ratios': [1.0, 31.0],
       'shots': None,
       'theta_intervals': [   [0, 0.25],
                              [0.08924749081234563, 0.09597799202522334],
                              [0.09214176853962333, 0.09235879737893671]]}

Code description
----------------

The code with primitives assumes that the user is running their code
locally, hence the reference to import
``from qiskit.primitives import Sampler``. After the algorithm is
adjusted to use a primitive, we initialize the primitive and then pass
it to the algorithm.

Step 1. Import the correct primitive package
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

How you import the primitive depends on what type of backend you want to run on.  

> * If you want to use **IBM Cloud resources** for simulating quantum circuits classically or running them on real quantum hardware, run ``from qiskit_ibm_runtime import Sampler``    ``from qiskit_ibm_runtime import QiskitRuntimeService``.

Reviewer: ```suggestion
* If you want to use IBM Cloud resources for simulating quantum circuits classically or **running them on real quantum hardware**, run ``from qiskit_ibm_runtime import Sampler``    ``from qiskit_ibm_runtime import QiskitRuntimeService``.
```
Feel like the real hardware is the bigger factor here

* If you don't need a lot of advanced features and **just want to run tests**, run ``from qiskit.primitives import Sampler``. This allows for an exact or shot-based classical simulation of quantum circuits. For the shot-based case, a normal probability distribution is fixed.
* If you want to run **advanced classical simulations** of quantum algorithms, run ``from qiskit.providers.aer.primitives import Sampler``.  This gives you access to an array of quantum circuit classical simulators that are better optimized and more customizable than the previous option.

.. note::
   
   Similar import options exist for the Estimator primitive. 

Step 2. Create the primitive instance
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To use default settings we use the following code:

``sampler = Sampler()``

We can modify run options which are passed to the primitive as
a Python dictionary. For example, setting the number of shots, which can
be done as follows:

.. code-block:: python

>    options = {"shots": 1024}

Reviewer: This doesn't work with runtime primitives. Qiskit Runtime primitives have more options, so the options are organized in categories. So to specify shots one would instead do

```
options  = {"execution": {"shots": 1024}}
```
(yes we should change it to make things easier for users)

   sampler = Sampler(options=options)

To learn about other options and the Sampler primitive in general, refer
to `Getting started with the sampler
primitive <https://qiskit.org/documentation/partners/qiskit_ibm_runtime/tutorials/how-to-getting-started-with-sampler.html>`__. 

Step 3. Use Sampler to initialize the algorithm and solve
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

After we set up the Sampler, we use it to initialize the Iterative
Amplitude Estimation algorithm and run the solve method with an
estimation problem instance, as follows:

.. code-block:: python

   problem = EstimationProblem(
       state_preparation=...,
       objective_qubits=[...],
       post_processing=...,
   )

   ae = IterativeAmplitudeEstimation(epsilon, alpha=alpha, sampler=sampler)

   result = ae.estimate(problem)

Step 4. Run the program
~~~~~~~~~~~~~~~~~~~~~~~

> To run the program in the cloud, we proceed as follows.

Reviewer: ```suggestion
To run the program using Qiskit Runtime primitive, we proceed as follows.
```

.. code-block:: python

   from qiskit_ibm_runtime import QiskitRuntimeService, Sampler, Session

   from qiskit.algorithms import IterativeAmplitudeEstimation, EstimationProblem
   from test.python.algorithms.test_amplitude_estimators import BernoulliStateIn, BernoulliGrover

> 
>    QiskitRuntimeService.save_account(

Reviewer: ```suggestion

   # This is only needed if account credentials was not already saved.
   QiskitRuntimeService.save_account(
```

       channel="ibm_cloud",
       token="",   # to be copied from the IBM Cloud account
       instance="crn:v1:bluemix:public:...",  # to be copied from the IBM Cloud account
       overwrite=True)

   service = QiskitRuntimeService()

   alpha = 0.05
   epsilon = 0.01
   shots = 4096
   prob = 0.3

   problem = EstimationProblem(BernoulliStateIn(prob), [0], BernoulliGrover(prob))

   with Session(service=service, backend="ibmq_qasm_simulator") as session:

>        sampler = Sampler(session=session, options={"shots": shots})

Reviewer: ```suggestion
       sampler = Sampler(session=session, options={"execution": {"shots": shots}})
```

Author: fixed
